backend/zkppSalt.js

```diff
@@ -110,35 +114,43 @@ export const registrationKey = async (contract: string, b: string): Promise<fals
   const encryptionKey = hashStringArray('REG', contract, registrationSecret).slice(0, nacl.secretbox.keyLength)
   const nonce = nacl.randomBytes(nacl.secretbox.nonceLength)
   const keyPair = boxKeyPair()
-  const s = Buffer.concat([nonce, nacl.secretbox(keyPair.secretKey, nonce, encryptionKey)]).toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')
+  const s = base64ToBase64url(Buffer.concat([nonce, nacl.secretbox(keyPair.secretKey, nonce, encryptionKey)]).toString('base64'))
   const now = (Date.now() / 1000 | 0).toString(16)
-  const sig = [now, Buffer.from(hashStringArray(contract, b, s, now, challengeSecret)).toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, '')].join(',')
+  const sig = [now, base64ToBase64url(Buffer.from(hashStringArray(contract, b, s, now, challengeSecret)).toString('base64'))].join(',')
 
   return {
     s,
-    p: Buffer.from(keyPair.publicKey).toString('base64').replace(/\//g, '_').replace(/\+/g, '-').replace(/=*$/, ''),
+    p: base64ToBase64url(Buffer.from(keyPair.publicKey).toString('base64')),
     sig
   }
 }
 
 export const register = async (contract: string, clientPublicKey: string, encryptedSecretKey: string, userSig: string, encryptedHashedPassword: string): Promise<boolean> => {
   if (!verifyChallenge(contract, clientPublicKey, encryptedSecretKey, userSig)) {
+    console.debug('register: Error validating challenge: ' + JSON.stringify({ contract, clientPublicKey, userSig }))
     return false
   }
 
   const record = await getZkppSaltRecord(contract)
 
   if (record) {
+    console.debug('register: Error obtaining ZKPP salt record for contract ID ' + contract)
```

Comment:
@corrideat Are these not errors?

> Seems like something that needs to be logged with `console.error`

There are several other places in this commit where `console.log` and `console.debug` were added instead of `console.error`, what's the reason for that?

Reply:
For consistency, this one should be an error. For the rest, as discussed on Slack, I feel that many of these 'errors' shouldn't be logged at the error level since they mean the protocol isn't executed correctly and are likely bad login attempts (which should be logged but are not necessarily an error in the application).

Reply:
If this literally where a bad login attempt would be detected? Or would that be detected elsewhere?

Bad login attempts can be logged with `console.warn`. We do want to log them, and also any attempts at hacking the protocol/server via bruteforce, so that tools like fail2ban can work automatically.